modulus/sym/geometry/geometry_dataloader.py

```diff
@@ -0,0 +1,292 @@
+        process_rank: int = 0,
+        world_size: int = 1,
+    ):
+
+        self.geom_objects = geom_objects
+        self.batch_size = batch_size
+        self.num_points = num_points
+        self.requested_vars = requested_vars
+        self.sample_type = sample_type
+        self.flip_interior = flip_interior
+        self.bounds = bounds
+        self.quasirandom = quasirandom
+        self.shuffle = shuffle
+        self.num_workers = num_workers
+        self.process_rank = process_rank
+        self.world_size = world_size
+
+        # Set up device, needed for pipeline
+        if isinstance(device, str):
+            device = torch.device(device)
+        # Need a index id if cuda
+        if device.type == "cuda" and device.index is None:
+            device = torch.device("cuda:0")
+        self.device = device
+
+        self.parse_dataset_files()
+
+        self.pipe = self._create_pipeline()
+
+        self.output_keys = self.requested_vars
+
+    def parse_dataset_files(self) -> None:
+        """
+        Parse the geometries.
+        """
+        # get the var names by sampling one of the geom
+        geo = self.geom_objects[0]
+        if self.sample_type == "surface":
+            samples = geo.sample_boundary(nr_points=self.num_points)
+        elif self.sample_type == "volume":
+            samples = geo.sample_interior(
+                nr_points=self.num_points, compute_sdf_derivatives=True
+            )
+
+        available_vars = list(samples.keys())
+        if self.requested_vars is None:
+            self.requested_vars = available_vars
+
+        if not set(self.requested_vars) <= set(available_vars):
+            raise ValueError(
+                f"Requested variables not available. Please select from {available_vars}"
+            )
+
+    def _create_pipeline(self) -> dali.Pipeline:
+        pipe = dali.Pipeline(
+            batch_size=self.batch_size,
+            num_threads=2,
+            prefetch_queue_depth=2,
+            py_num_workers=self.num_workers,
+            device_id=self.device.index,
+            py_start_method="spawn",
+        )
+
+        with pipe:
+            source = GeometrySource(
+                geom_objects=self.geom_objects,
+                batch_size=self.batch_size,
+                num_points=self.num_points,
+                requested_vars=self.requested_vars,
+                sample_type=self.sample_type,
+                bounds=self.bounds,
+                quasirandom=self.quasirandom,
+                shuffle=self.shuffle,
+                process_rank=self.process_rank,
+                world_size=self.world_size,
+            )
+
+            self.length = len(source) // self.batch_size
+            vars_tuple = dali.fn.external_source(
+                source,
+                num_outputs=len(self.requested_vars),
+                parallel=False,  # Parallel=False to simplify the code, you can switch to True if your source is thread-safe
```

Comment:
I suggest exposing this through the datapipe API

Reply:
Good catch. fixed